### ISSUE TITLE
Development and production environments

### ISSUE THREAD

**Issue by NanneHuiges** [maximebf/php-debugbar#435]
What is the preferred way to go about having a 'clean' production environment? 

It might be so trivial that it's not mentioned and/or I'm just overlooking the hints in the docs, but I don't see much mention about environments. I would rather not have the debug-code running on production. The manual implies you use a normal `require` and not `require-dev`, so if I follow that you'd get the code online at least. THis is apart from the fact that you will have calls on several places?

You can probably easily not-render the bar in a production environment, but then, correct me if I'm wrong, your collectors are still collecting, using resources and possibly affect stuff. I suppose most of the collection will not be in a single place and therefore I'm not seeing a quick method to not-have that code running on a production environment. Only thing I could come up with is switching out your collectors with injection for some sort of dummy / empty collector that does nothing. But that feels hacky, and as it's not mentioned anywhere that's probably not the default method :)

Is there a best-practice somewhere to be found?

**Comment by oscarlosan** [comment]
I am looking for the same. There is no point in using this library if we can't disable it in production.

**Comment by hanleybrand** [comment]
This is an old issue, and I think someone else mentioned a way to do it via composer in another issue (i.e. don't install it in prod)-- but here's a quick and dirty solution I've used before (I edited a file on my HD, apologies if the formatting is weird or anything is out of date, but this should work). 

On reflection, it could be refactored to check if DebugBar is available (i.e. with `if (class_exists('StandardDebugBar'){}` or similar) and to only work when it's installed instead of checking prod/non-prod.

Regardless, the main value of implementing the helper methods in `dbgbr.php` below (whether they're gated by env values or the existence of the StandardDebugBar class) is not having to strip out your debugbar calls for production stages, as well as having a single statement can do double time for error logging when desired.

1st create `dbgbr.php` somewhere

```php
<?php
require_once  'vendor/autoload.php';

use DebugBar\StandardDebugBar;
$debugbar = new StandardDebugBar();
$debugbarRenderer = $debugbar->getJavascriptRenderer();

/** set an env STAGE_ID= prod||non-prod on your servers. */
function is_not_prod(): bool { return (getenv('STAGE_ID') != 'prod'); }
$is_not_prod = is_not_prod();

/** array of users you are willing to allow to use the tool - you could also probably do an IP based allow list, etc  */
$user_allow_list = array('dev1', 'dev2');

/** Sends $message to the phpdebugbar when in use (non-prod only)
 * @param string|array|object $message
 * @param bool|string $always_log -- true for string argument 'true' otherwise false
 *
 *        use string 'true' as the second argument to have the message written to the error logs.
 *        Note: $message of type array|object tend to clutter the error log, so use sparingly
 *
 * @return void
 */
function dbb_msg(string|array|object $message, bool|string $always_log = null): void
{
	global $debugbar, $is_not_prod;
	// set $always_log false unless true or 'true'
	if (!isset($always_log)) {
		$always_log = false;
	}
	if (is_String($always_log)) {
		$always_log = $always_log == "true";
	}
	if ($message && isset($debugbar) && $is_not_prod) {
		$debugbar['messages']->info($message);
	}
	if ($always_log) {
                // this should work with any value you that works with $debugbar['messages']
                $file = basename($_SERVER["SCRIPT_FILENAME"]);
		$message = is_string($message) ? "$file : $message" : $message;
		ob_start();                                                       // start buffer capture
		var_dump($message);                                   // dump the values
		$contents = ob_get_contents();                    // put the buffer into a variable
		ob_end_clean();                                             // end capture
		error_log(var_export($contents, true));        // log contents of the result of var_dump( $message )
	}
}

function echo_debug_bar_scripts(): void {
	global $is_not_prod, $user_allow_list;
	$session_username = $_SESSION['username'] ?? false;   // or however you deal with the logged in user's  username
       /** simpler, less secure, but also allows debugging when not logged in:
        * if ( $is_not_prod)
        */
	if ($session_username && in_array($session_username, array($user_allow_list)) && $is_not_prod) {
		global $debugbarRenderer;
		echo $debugbarRenderer->renderHead(); /** still works if not in <head> */
		echo $debugbarRenderer->render();
		// optional
                //echo "<script>console.log('Debugbar activated for $session_username')</script>";
	} else {
		echo "<script>console.log('Hi $session_username, this app was made with ❤️ by us for you')</script>";
	}
}
$inc = get_included_files();
if (basename(__FILE__) == basename($inc[0])) exit();
```
then in any file you want to use the debugbar with, add this:

```php
<?php

require_once("{$path_to}/dbgbr.php");

dbb_msg("a detail of logic flow or page content useful during development" , 'false');

if (!$sql_stmt) {dbb_msg(print_r(sqlsrv_errors(), true), "true"); } //always log sql errors

// you can use the  is_not_prod() function to only log dbb_msg statements in the error_log on prod
$log_msg = is_not_prod() ? 'false' : 'true'; 
// imagine you did some api call or whatever using $result for the response
dbb_msg(["api result"=>$result], $log_msg);

//  this needs to execute somewhere in the <body> tag
echo_debug_bar_scripts();
// I usually do it with a template that lacks the last two salient tags:
echo "</body>\n</html>";

```

Hope this is helpful

**Comment by parallels999** [comment]
I made it more automatic on a bootstrap file
```php
if(class_exists(\DebugBar\DebugBar::class) && in_array($ses_username, array($user_allow_list)) && $is_not_prod) {
    class DebugBar extends \DebugHelper {}
    DebugBar::initialize(__DIR__.'/../../profiles', __DIR__.'/open.php', __DIR__.'/../vendor/src/DebugBar/Resources');
    DebugBar::setEditor('vscode', '/xampp/htdocs/app');
    //DebugBar::enableFileTraces(false); // avoid php bactraces, faster
    DebugBar::enableGlobalTimeline(true); // timeline on all collectord
    DebugBar::collectWarnings(false); // false avoid php default error_handler, true let it
    //DebugBar::log(array('test' => 'test')); 
    //DebugBar::addException(new \Exception('Exception'));
    //DebugBar::setDebugBar(null); // disable debugbar
}else{
    class DebugBar {
        public static function __callStatic($name, $arguments) {
            if($name == 'measure' && isset($arguments[1]) && is_callable($arguments[1]))
                $arguments[1](); return null;
        } 
    }
}
```
`DebugHelper` ia a class, which handles the debugbar methods statically, and if it goes through the `else` it returns null, so the code `\DebugBar::log('test');` can remain and does not affect production.
and for the problem of injecting the debugbar into the html what I do is in my method `DebugBar::initialize` add it automatically to every call with `ob_start` and `register_shutdown_function`

```php
ob_start();

register_shutdown_function(function () {
    if (DebugHelper::isAjax() || !DebugHelper::hasDebugBar()){
        if(!DebugHelper::stacked())
            DebugHelper::sendDataInHeaders(true);
        ob_end_flush();
    }else {
        $content = ob_get_contents();
        ob_get_clean(); // Clean buffer

        if (DebugHelper::stacked()) {
            echo $content;
            return;
        }

        // Try to put the js/css directly before the </head>
        $pos = stripos($content, '</head>');
        if (false !== $pos) {
            $content = substr($content, 0, $pos) . DebugHelper::renderHead() . substr($content, $pos);

            // Try to put the widget at the end, directly before the </body>
            $pos = strripos($content, '</body>');
            if (false !== $pos) {
                $content = substr($content, 0, $pos) . DebugHelper::render() . substr($content, $pos);
            } else {
                DebugHelper::addWarning(E_USER_NOTICE, '`</body>` tag not found on html string.');
                $content = $content . DebugHelper::render();
            }
        } else {
            DebugHelper::addWarning(E_USER_NOTICE, '`</head>` tag not found on html string.');
            DebugHelper::getData();
        }

        echo $content; // show the content
    }
});
```

This is how I handle it automatically without having to remove my debugbar code and without worrying about adding the renders. I don't know if it's the most optimal, but it's been working for me for a long time.

```php
class DebugHelper
{
    /**
     * Calls are automatically redirected to their specific collector in `__callStatic` (if collector exists)
     * Similar to https://github.com/barryvdh/laravel-debugbar/blob/master/src/LaravelDebugbar.php, but more automatic
     * 
     * @method static void addException(\Exception $e)
     * @method static void addThrowable(\Throwable $t)
     * @method static void addWarning(int $errno, string $errstr, string $errfile = '', int $errline = 0)
     * @method static array collectWarnings(bool $preserveOriginalHandler = false)
     *
     * @method static void addMeasure(string $label, float $start, float $end, array $params = [], string $collector = null, string $group = null)
     * @method static mixed measure(string $label, callable $callback, string $collector = null)
     * @method static void startMeasure(string $name, string $label, string $collector = null, string $group = null)
     * @method static void stopMeasure(string $name, array $params = [])
     * @method static bool hasStartedMeasure(string $name)
     *
     * @method static void addMessage(mixed $message, string $label = 'info', mixed $isString = true)
     * @method static void aggregate(\DebugBar\DataCollector\MessagesAggregateInterface $messages)
     * @method static void emergency(string $message, array $context = array())
     * @method static void alert(mixed $message, array $context = array())
     * @method static void critical(mixed $message, array $context = array())
     * @method static void error(mixed $message, array $context = array())
     * @method static void warning(mixed $message, array $context = array())
     * @method static void notice(mixed $message, array $context = array())
     * @method static void info(mixed $message, array $context = array())
     * @method static void debug(mixed $message, array $context = array())
     *
     * @method static void countClass(mixed $class, int $count = 1)
     *
     * @method static void startQueryMeasure()
     * @method static void addQuery(string $query, array $extraData = array())
     * @method static void addTransactionEvent(string $event, array $extraData = array())
     * @method static void addComment(string $comment, array $extraData = array())
     */
    //......
